lms/djangoapps/lti_provider/users.py

```diff
@@ -28,14 +28,35 @@ def authenticate_lti_user(request, lti_user_id, lti_consumer):
     If the currently logged-in user does not match the user specified by the LTI
     launch, log out the old user and log in the LTI identity.
     """
+    lis_email = request.POST.get("lis_person_contact_email_primary")
+
+    # Verify that the email from the LTI Launch and the logged-in user are the same.
+    if lti_consumer.auto_link_users_using_email and (
+        not request.user.is_authenticated or
```

Comment:
> The auto-linking is done only when request.user.email and the lis_person_contact_email_primary match. **If the user is not-authenticated** or authenticated with a user account with different email, the session will be cleared, and an error message is shown. 

This is different from the initial approach. Do we want to force users to create an account when using this option? If yes, perhaps we should create a choice field with "the default approach", "auto-linking", and "auto-linking with registration requirement" (we should think about better names for these). We could also simply change the name of this field to indicate the implications.

Reply:
Yes, we would like to make sure all the users have an account on Open edX before they access the content via LTI. This is an intentional change to maintain consistency. With the earlier implementation, users with and without an account had different experiences. Now everyone has the same experience and they are all expected to have an account.

To reflect this, I have renamed the flag to `require_user_account` and added help text explaining the outcome of enabling this flag. The [documentation has been updated](https://github.com/openedx/edx-documentation/pull/2190/files) with the relevant details as well.

Reply:
@Agrendalath Hmm. I thought requiring that the Open edX user being logged in made it avoid the security issue of spoofed emails. So I didn't see the need to handle the user being logged out. 

I am also wondering how to implement the email validation when we don't have an `request.user`.  Perhaps use the `lti_user_id` which is supposed to be unique and then throw an error if the `lti_user.edx_user.email` doesn't match the `lis_email`? 🤔 But that would also mean, sometimes the `edx_user` could be the anonymous account (flag turned on after a learner has used it anonymously) and their email would never match. Then we will have to filter for the `lti.example.com` in the email to identify anonymous vs real accounts. I wonder if this level checking is worth it.

Is requiring the users being logged in to both the systems a bad requirement?

Reply:
@tecoholic,
> Is requiring the users being logged in to both the systems a bad requirement?

I believe so. One of the major benefits of using Open edX as an LTI provider is the seamless integration with another LMS. Otherwise, we could add an annotation that you need to be logged in to access the content (provide a link to the Open edX login page), and just embed the `/xblock/{block_id}` view in an iframe below.

> Hmm. I thought requiring that the Open edX user being logged in made it avoid the security issue of spoofed emails.

Not really - we care about this issue only when linking a new account My initial explanation about the vulnerability might have been a bit fuzzy. Let's compare this to the OAuth2 mechanism, which is very similar in this aspect.

It consists of the SP (our instance) and the IdP (external service). Let's say we use site A as the IdP (_I wanted to call it X, but it could be controversial, given the recent branding change of one of the bigger sites_). If we fully trust A before linking the user account, they can provide any email to take over any account (e.g., one with the superuser permissions) on our instance. Therefore, we don't want to fully trust them with linking the users. Instead, it's the user who should decide if they want to trust this site to act as an IdP on their behalf. When they link the account, they can log into the platform without a password when they are authenticated on A. However, if A decides to provide a fake email address, they will not be able to access the superuser who does not have the account linked with their service. Therefore, _we_ are not trusting them with our whole user base - instead, it's the _individual users_ who trust them with their accounts.

Back to the LTI - since the `LtiUser` is tied to a specific `LtiConsumer`, there is no way for the LTI consumer to take over an account of a user who did not link the account with this service first. This way, we limit the attack surface to users who decide to use their system, without sacrificing the major benefit of using another LMS as the seamless IdP.

> Perhaps use the lti_user_id which is supposed to be unique (...)

Yes - [it is unique](https://github.com/openedx/edx-platform/blob/390c2fb8ca0ef068d199587cf302551f8d2459ab/lms/djangoapps/lti_provider/models.py#L151-L152) for the `lti_consumer` + `lti_user_id` pair.

> I am also wondering how to implement the email validation when we don't have an request.user.

We don't need to check the email. We already get the `LtiUser` [here](https://github.com/openedx/edx-platform/blob/390c2fb8ca0ef068d199587cf302551f8d2459ab/lms/djangoapps/lti_provider/users.py#L41-L44), based on the `lti_user_id` ([retrieved from the LTI launch params](https://github.com/openedx/edx-platform/blob/5b464b6a0b078c0a55c6fba33276c22d74fde588/lms/djangoapps/lti_provider/views.py#L94)) and `lti_consumer` ([retrieved from our DB, with the GUID/key validation](https://github.com/openedx/edx-platform/blob/5b464b6a0b078c0a55c6fba33276c22d74fde588/lms/djangoapps/lti_provider/views.py#L66-L69)).

> and then throw an error if the lti_user.edx_user.email doesn't match the lis_email? 🤔

We don't care about it at this point. The `LtiUser` already exists, which means that the user has already linked their account.

> But that would also mean, sometimes the edx_user could be the anonymous account (flag turned on after a learner has used it anonymously) and their email would never match. Then we will have to filter for the lti.example.com in the email to identify anonymous vs real accounts. I wonder if this level checking is worth it.

As we don't care about the email for the already linked accounts, we do not need to handle this scenario.

Reply:
@Agrendalath Thank you for explaining everything in detail. The explanation on balancing trust between the entities was really helpful. 

I will work on the changes to handle the missing scenario and ping you again for review.

Reply:
@Agrendalath I have fixed the issue and tested that LTI launch will succeed after the user is logged out. I also added a test case to verify this in 78c5b9c47bcf2a2f1e9eec54e5404fc69c521d86.